docs/proposals/010-new-diff-type-system.md

```diff
@@ -0,0 +1,283 @@
+2.  `DiffTensorView<T>` always has two buffer pointers even when used in non-differentiable methods. This is extra data in the struct, and potentially extra tensor allocations (we explicitly handle this case in `slangtorch` by leaving the diff part uninitialized if a primal method is invoked)
+3.  Higher-order derivatives don't work well with this workaround. Differentiating a method twice needs a set of 4 pointers, but we need to account for this ahead of time by using new types like `DiffDiffTensorView` that worsens the problem of carrying around extra data where its not required.
+
+
+## Solution
+
+We'll need to make the following 4 additions/changes:
+### 1. `[deriv_method]` function decorator.
+Intended for easy definition of custom derivatives for struct methods. It has the following properties:
+1. Accesses to `this` within `[deriv_method]`  are differential pairs.
+2. Methods decorated with `[deriv_method]` cannot be called as regular methods (they can still be explicitly invoked with `bwd_diff(obj.method)`), and do not show up in the auto-complete list.
+
+See the next section for example uses of `[deriv_method]`.
+
+### 2. Split `IDifferentiable` interface: `IDifferentiableValueType` and `IDifferentiablePtrType`
+This approach moves away from "type-driven" derivative semantics and towards more "function-driven" derivative semantics.
+We no longer have a `dadd` , `dzero`, `dmul` etc.. we use default initialization instead of `dzero` and the backward derivative of the `use` method for `dadd`
+
+Further, `IDifferentiablePtrType` types don't have any of these properties. They do not need a way to 'add', and it is especially important that there is no default initializer. We never want the compiler to be able to create a new object of  `IDifferentiablePtrType` since we want to get the user-provided pointers.
+
+Additionally, we can use `IDifferentiableValueType` as the current `IDifferentiable` for backwards compatibility (it should just work in 95% of cases, since no one really defines dadd/dzero/dmul explicitly anyway) 
+
+Here's the new set of base interfaces:
+```csharp
+interface __IDifferentiableBase { } // Helper type for our implementation.
+interface IDifferentiableValueType : __IDifferentiableBase
+{
+	associatedtype Differential : IDifferentiableValueType & IDefaultInitializable;
+	[Differentiable] This use(); // auto-synthesized
+}
+
+interface IDifferentiablePtrType : __IDifferentiableBase
+{
+	associatedtype Differential : IDifferentiablePtrType;
+}
+
+```
+
+Some extras in stdlib allow us to constrain the diffpair type for things like `IArithmetic`
+```csharp
+// --- STDLIB EXTRAS ---
+
+interface ISelfDifferentiableValueType : IDifferentiableValueType
+{
+	// Force arithmetic types to be a differential pair of the same two types.
+	// Make it simple to define derivatives of arithmetic operations.
+	//
+	associatedtype Differential : This;
```

Comment:
It is about time we add a `where` clause and implement type equality constraints.

Reply:
Yes, this part is confusing.. what we want is the equality constraints. 
(We could implement `ISelfDifferentiableValueType` as a special case for now, in case the `where` clause implementation is too complex)